### DISCUSSION TITLE
Carrier Billing API: Treating x-correlator header field Mandatory for a specific Use-case

### DISCUSSION THREAD

**Discussion by ajmhussain** [camaraproject/Commonalities#287] [Q&A] [answered]
As per Camara specifications for Carrier Billing APIs, x-correlator as a header field is defined as optional. But in order to correlate the requests end to end for our critical business use-case, we have a requirement to enforce it so that we always get a correlation id from the clients and use it to track end-to-end. Can we use an optional header field as mandatory as per our requirement keeping our APIs in compliance to Camara standards?

**Comment by hdamker** [comment, marked as answer]
@ajmhussain The requirements regarding the x-correlator header are described within the [chapter 9 of the CAMARA Design Guidelines](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#9-architecture-headers), stating that the header parameter is not required by the API Consumer. And:

> When the API Consumer includes the "x-correlator" header in the request, the API provider must include it in the response with the same value that was used in the request. Otherwise, it is optional to include the "x-correlator" header in the response with any valid value. Recommendation is to use UUID for values.

So you can use the header in very request, either with the value provided by the API Consumer or by setting the value in the response yourself.
